src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CredScanSuppressionsGenerator.cs

```diff
@@ -0,0 +1,204 @@
+}
+
+public class CredScanSuppressionsGenerator : ICredScanSuppressionsGenerator
+{
+    private static readonly IReadOnlyCollection<LocalPath> s_credScanSuppressionsLocations = new[]
+    {
+        new UnixPath(".config/" + VmrInfo.CredScanSuppressionsFileName),
+        new UnixPath("eng/" + VmrInfo.CredScanSuppressionsFileName),
+    };
+
+    private readonly IVmrInfo _vmrInfo;
+    private readonly ISourceManifest _sourceManifest;
+    private readonly ILocalGitClient _localGitClient;
+    private readonly IFileSystem _fileSystem;
+    private readonly ILogger<CredScanSuppressionsGenerator> _logger;
+    private readonly JsonSerializerOptions _jsonOptions = new()
+    {
+        AllowTrailingCommas = true,
+        ReadCommentHandling = JsonCommentHandling.Skip,
+        WriteIndented = true,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+    };
+
+    public CredScanSuppressionsGenerator(
+        IVmrInfo vmrInfo,
+        ISourceManifest sourceManifest,
+        ILocalGitClient localGitClient,
+        IFileSystem fileSystem,
+        ILogger<CredScanSuppressionsGenerator> logger)
+    {
+        _vmrInfo = vmrInfo;
+        _sourceManifest = sourceManifest;
+        _localGitClient = localGitClient;
+        _fileSystem = fileSystem;
+        _logger = logger;
+    }
+
+    /// <summary>
+    /// Generates the CredScanSuppressions.json file by gathering individual repo CredScanSuppressions.json files.
+    /// </summary>
+    public async Task UpdateCredScanSuppressions(CancellationToken cancellationToken)
+    {
+        _logger.LogInformation("Updating {credscansuppressions}...", VmrInfo.CredScanSuppressionsPath);
+
+        var destPath = _vmrInfo.VmrPath / VmrInfo.CredScanSuppressionsPath;
+
+        CredScanSuppressionFile vmrCredScanSuppressionsFile = new CredScanSuppressionFile();
+
+        _fileSystem.CreateDirectory(_fileSystem.GetDirectoryName(destPath)
+            ?? throw new Exception($"Failed to create {VmrInfo.CredScanSuppressionsFileName} in {destPath}"));
+
+        using (var destStream = _fileSystem.GetFileStream(destPath, FileMode.Create, FileAccess.Write))
+        {
+            foreach (ISourceComponent component in _sourceManifest.Repositories.OrderBy(m => m.Path))
+            {
+                await AddCredScanSuppressionsContent(vmrCredScanSuppressionsFile, component.Path, cancellationToken);
+
+                foreach (var submodule in _sourceManifest.Submodules.Where(s => s.Path.StartsWith($"{component.Path}/")))
+                {
+                    await AddCredScanSuppressionsContent(vmrCredScanSuppressionsFile, submodule.Path, cancellationToken);
+                }
+            }
+
+            JsonSerializer.Serialize(destStream, vmrCredScanSuppressionsFile, _jsonOptions);
+        }
+
+        if (vmrCredScanSuppressionsFile.Suppressions.Count == 0)
+        {
+            _fileSystem.DeleteFile(destPath);
```

Comment:
If we deleted the file, should we stage that change?

Reply:
Should we also test existence?

Reply:
> If we deleted the file, should we stage that change?

So if we remove the last entry it gets deleted? I assume so, but we should probably do the same here then: https://github.com/dotnet/arcade-services/blob/9d6a955597b3e5081a2ad82dc0d80eaf9f3359c7/src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CodeownersGenerator.cs#L97-L104

> Should we also test existence?

I don't think so, the intention is to overwrite the file.

Reply:
Ah, no gdn files will still produce an empty file that will overwrite whatever was or was not there.

But yeah, if we remove all gdn files from repos, the file will stay in the VMR in its last state so trying to stage the removal might be worth doing. Just not sure if stage of a non-change fails or no-ops successfully.

Reply:
I tried it with `git` and staging a file without changes works fine.

Reply:
ah, but if we don't have any entries and never add the file staging a non-existing file would fail. so we need to check whether the file existed before